ingestion-beam/src/main/java/com/mozilla/telemetry/util/KeyStore.java

```diff
@@ -0,0 +1,122 @@
+    ensureKeysLoaded();
+    PrivateKey key = keys.get(path);
+    return key;
+  }
+
+  private final String metadataLocation;
+  private final boolean kmsEnabled;
+  private transient Map<String, PrivateKey> keys;
+
+  public static KeyStore of(String metadataLocation, boolean kmsEnabled) {
+    return new KeyStore(metadataLocation, kmsEnabled);
+  }
+
+  private KeyStore(String metadataLocation, boolean kmsEnabled) {
+    this.metadataLocation = metadataLocation;
+    this.kmsEnabled = kmsEnabled;
+  }
+
+  @VisibleForTesting
+  int numLoadedKeys() {
+    ensureKeysLoaded();
+    return keys.size();
+  }
+
+  private void loadAllKeys() throws IOException, ValidationException {
+    final Map<String, PrivateKey> tempKeys = new HashMap<>();
+
+    Schema schema;
+    try (InputStream inputStream = Resources.getResource("keystore-metadata.schema.json")
+        .openStream()) {
+      schema = SchemaLoader.load(Json.readJsonObject(inputStream));
+    } catch (IOException e) {
+      throw new IOException("Exception thrown while reading metadata file", e);
+    }
+
+    // required to validate Jackson objects
+    JsonValidator validator = new JsonValidator();
+
+    ArrayNode metadata;
+    try (InputStream inputStream = BeamFileInputStream.open(this.metadataLocation)) {
+      byte[] data = IOUtils.toByteArray(inputStream);
+      metadata = Json.readArrayNode(data);
+      validator.validate(schema, metadata);
+    } catch (IOException e) {
+      throw new IOException("Exception thrown while reading keystore metadata schema.", e);
+    }
+
+    for (JsonNode element : metadata) {
+      String privateKeyId = element.get("private_key_id").textValue();
+      String privateKeyUri = element.get("private_key_uri").textValue();
+      String kmsResourceId = element.get("kms_resource_id").textValue();
+
+      try (InputStream inputStream = BeamFileInputStream.open(privateKeyUri)) {
+        byte[] keyData = IOUtils.toByteArray(inputStream);
+
+        PublicJsonWebKey key;
+        if (kmsEnabled) {
+          try (KeyManagementServiceClient client = KeyManagementServiceClient.create()) {
+            DecryptResponse response = client.decrypt(kmsResourceId, ByteString.copyFrom(keyData));
+            key = PublicJsonWebKey.Factory
+                .newPublicJwk(new String(response.getPlaintext().toByteArray()));
+          }
+        } else {
+          key = PublicJsonWebKey.Factory.newPublicJwk(new String(keyData));
+        }
+
+        tempKeys.put(privateKeyId, key.getPrivateKey());
+      } catch (IOException e) {
+        throw new IOException("Exception thrown while reading key specified by metadata.", e);
+      } catch (JoseException e) {
+        throw new RuntimeException(e);
+      }
+    }
+
+    keys = tempKeys;
+  }
+
+  private void ensureKeysLoaded() {
+    if (keys == null) {
+      try {
+        loadAllKeys();
+      } catch (IOException e) {
+        throw new UncheckedIOException("unexpected error while loading keys", e);
```

Comment:
It looks like we already explicitly catch and wrap all exception cases in `loadAllKeys`. I think I'd prefer to make sure those are all runtime exceptions (UncheckedIOException in particular where it makes sense) with helpful explanation messages and avoid an additional layer in the stack trace here.

Reply:
I'll leave out the message here, but wrap the IOExceptions in the unchecked wrapper.